### ISSUE TITLE
--with-default-msvcrt=ucrt breaks libgomp build

### ISSUE THREAD

**Issue by Prince213** [niXman/mingw-builds#520]
Command: `./build --mode=gcc-10.2.0 --buildroot=/e/mingw1020 --no-multilib --no-extras --jobs=6 --rt-version=v7.0.0 --with-default-msvcrt=ucrt --enable-languages=c,c++ --threads=posix --exceptions=seh --arch=x86_64 --bin-compress`
Log:
```
../../../../../src/gcc-10.2.0/libgomp/env.c: In function 'handle_omp_display_env':
../../../../../src/gcc-10.2.0/libgomp/env.c:1285:24: error: unknown conversion type character 'l' in format [-Werror=format=]
 1285 |       fprintf (stderr, "  GOMP_SPINCOUNT = '%"PRIu64"'\n",
      |                        ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../../../src/gcc-10.2.0/libgomp/env.c:40:
E:/mingw1020/x86_64-1020-posix-seh-rt_v700/mingw64/mingw/include/inttypes.h:36:18: note: format string is defined here
   36 | #define PRIu64 "llu"
      |                  ^
../../../../../src/gcc-10.2.0/libgomp/env.c:1285:24: error: too many arguments for format [-Werror=format-extra-args]
 1285 |       fprintf (stderr, "  GOMP_SPINCOUNT = '%"PRIu64"'\n",
      |                        ^~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by niXman** [comment]
https://stackoverflow.com/a/55099671

**Comment by Prince213** [comment]
Well, I think the problem is caused by _UCRT...

**Comment by Prince213** [comment]
```c
/* 7.8.1 Macros for format specifiers
 * 
 * MS runtime does not yet understand C9x standard "ll"
 * length specifier. It appears to treat "ll" as "l".
 * The non-standard I64 length specifier causes warning in GCC,
 * but understood by MS runtime functions.
 */
#if defined(_UCRT) || __USE_MINGW_ANSI_STDIO
#define PRId64 "lld"
#define PRIi64 "lli"
#define PRIo64 "llo"
#define PRIu64 "llu"
#define PRIx64 "llx"
#define PRIX64 "llX"
#else
#define PRId64 "I64d"
#define PRIi64 "I64i"
#define PRIo64 "I64o"
#define PRIu64 "I64u"
#define PRIx64 "I64x"
#define PRIX64 "I64X"
#endif
```

**Comment by niXman** [comment]
I think it should be `#if defined(_UCRT) && !defined(__USE_MINGW_ANSI_STDIO)`, right?

**Comment by Prince213** [comment]
IDK :( . the second answer to the Stack Overflow question says:
> If you must use %llu, define __USE_MINGW_ANSI_STDIO macro before including stdio.h. Be aware that if you do this, MS type %I64* format will no longer work.

**Comment by niXman** [comment]
ah right

**Comment by niXman** [comment]
@Prince213 look at this: https://sourceforge.net/p/mingw-w64/bugs/853/

**Comment by Biswa96** [comment]
Try with this patch https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-gcc/0020-libgomp-Don-t-hard-code-MS-printf-attributes.patch

**Comment by Prince213** [comment]
Errors related to env.c now disappears, but new ones come:
```
In file included from ../../../../../src/gcc-10.2.0/libgomp/config/mingw32/affinity-fmt.c:68:
E:/mingw1020/src/gcc-10.2.0/libgomp/affinity-fmt.c: In function 'gomp_display_affinity':
E:/mingw1020/src/gcc-10.2.0/libgomp/affinity-fmt.c:388:22: error: unknown conversion type character 'l' in format [-Werror=format=]
  388 |        sprintf (buf, "0x%" PRIx64, (uint64_t) gomp_integral (handle));
      |                      ^~~~~
In file included from ../../../../../src/gcc-10.2.0/libgomp/config/mingw32/affinity-fmt.c:34:
E:/mingw1020/x86_64-1020-posix-seh-rt_v800/mingw64/mingw/include/inttypes.h:37:18: note: format string is defined here
   37 | #define PRIx64 "llx"
      |                  ^
In file included from ../../../../../src/gcc-10.2.0/libgomp/config/mingw32/affinity-fmt.c:68:
E:/mingw1020/src/gcc-10.2.0/libgomp/affinity-fmt.c:388:22: error: too many arguments for format [-Werror=format-extra-args]
  388 |        sprintf (buf, "0x%" PRIx64, (uint64_t) gomp_integral (handle));
      |                      ^~~~~
```

**Comment by lurobi** [comment]
Just some extra information: I think I ran into a related issue.  I got: `unknown conversion type character 'l' in format"` in `gomp_fatal` 

My build was: `./build --no-extras --mode=gcc-9.3.0 --arch=x86_64 --exceptions=seh --buildroot=....` (From mingw-builds:master) and it would not complete until I applied the patch noted by @Biswa96

**Comment by niXman** [comment]
@lurobi  please try `develop` branch

**Comment by niXman** [comment]
@Prince213 any ideas?

**Comment by Prince213** [comment]
@lurobi @niXman 7de159b732fd2f428048f0574a9c9124dd827774 should fix it, which is already on develop.

**Comment by niXman** [comment]
@Prince213 thank you!

@lurobi can you confirm please?

**Comment by lurobi** [comment]
Confirmed: I was able to build gcc 9 and 10 branches on develop because of
that patch.  I hit an unrelated issue which prevented me from compiling gcc
trunk.

I am not specifically setting that msvcrt flag if that is relevant.

**Comment by Prince213** [comment]
@lurobi Please try with --with-default-msvcrt=ucrt as this problem is probably related to UCRT.

**Comment by lurobi** [comment]
I am unable to compile gcc using `--with-default-msvcrt=ucrt` on develop on windows through MSYS2.  I am using a1c154f2 plus my fix for the binutils heredoc.

Command: `./build --mode=gcc-10.2.0 --buildroot=/home/luker/mingw1020 --no-multilib --no-extras --jobs=6 --rt-version=v7.0.0 --with-default-msvcrt=ucrt --enable-languages=c,c++,fortran --threads=posix --exceptions=seh --arch=x86_64 --bin-compress`

```
../../../../../src/gcc-10.2.0/libgomp/env.c: In function 'handle_omp_display_env':
../../../../../src/gcc-10.2.0/libgomp/env.c:1285:24: error: unknown conversion type character 'l' in format [-Werror=format=]
 1285 |       fprintf (stderr, "  GOMP_SPINCOUNT = '%"PRIu64"'\n",
      |                        ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../../../src/gcc-10.2.0/libgomp/env.c:40:
C:/msys64/home/luker/mingw1020/x86_64-1020-posix-seh-rt_v700/mingw64/mingw/include/inttypes.h:36:18: note: format string is defined here
   36 | #define PRIu64 "llu"
      |                  ^
../../../../../src/gcc-10.2.0/libgomp/env.c:1285:24: error: too many arguments for format [-Werror=format-extra-args]
 1285 |       fprintf (stderr, "  GOMP_SPINCOUNT = '%"PRIu64"'\n",
      |                        ^~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by Biswa96** [comment]
To disable `-Werror` for UCRT GCC build, some workaround are added in this commit https://github.com/msys2/MINGW-packages/commit/5d82e17a56a3216a54dbb5924b92c96284ae5c97

**Comment by lazka** [comment]
Here is the upstream fix: https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=c51f1e7427e6a5ae2a6d82b5a790df77a3adc99a

**Comment by sfan5** [comment]
Seeing the same errors with gcc 10.3.0 and UCRT:
<pre>
In file included from /home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/affinity-fmt.c:68:
/home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/../../affinity-fmt.c: In function 'gomp_display_affinity':
/home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/../../affinity-fmt.c:388:22: error: unknown conversion type character 'l' in format [-Werror=format=]
  388 |        sprintf (buf, "0x%" PRIx64, (uint64_t) gomp_integral (handle));
      |                      ^~~~~
In file included from /home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/affinity-fmt.c:34:
/usr/i686-w64-mingw32/include/inttypes.h:37:18: note: format string is defined here
   37 | #define PRIx64 "llx"
      |                  ^
In file included from /home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/affinity-fmt.c:68:
/home/build/mingw-w64-gcc/gcc/libgomp/config/mingw32/../../affinity-fmt.c:388:22: error: too many arguments for format [-Werror=format-extra-args]
  388 |        sprintf (buf, "0x%" PRIx64, (uint64_t) gomp_integral (handle));
      |                      ^~~~~
</pre>

Looking at `gcc-10.3.0/gcc/config/i386/msformat-c.c` I don't see the changes that should be there.
So it seems that the fix hasn't made it into gcc 10.3. Anyone know why?

**Comment by jrmoore** [comment]
Seeing the same error over here when not using `--static-gcc` trying to build with the following command line:

```
./build --jobs=1 --mode=gcc-9.4.0 --rt-version=v8.0.0 --with-default-msvcrt=ucrt --arch=x86_64 --buildroot=/builds \
--exceptions=seh --threads=win32 --enable-languages=c,c++ --no-multilib --no-extras
```

It seems another patch is necessary for that configuration to build, not that I'm interested in it, I'm trying to find the culprit for #498 and stumbled upon this, right now I can't find a way to build a toolchain using UCRT.
